src/main/java/com/hrznstudio/titanium/api/multiblock/MultiblockTemplate.java

```diff
@@ -0,0 +1,156 @@
+package com.hrznstudio.titanium.api.multiblock;
+
+import com.google.common.collect.ImmutableList;
+import com.hrznstudio.titanium.api.multiblock.matcher.IMatch;
+import com.hrznstudio.titanium.api.multiblock.matcher.Matcher;
+import com.hrznstudio.titanium.util.BlockPosUtil;
+import com.hrznstudio.titanium.util.RotationUtil;
+import com.hrznstudio.titanium.util.StaticTemplateUtil;
+import net.minecraft.block.BlockState;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.util.Direction;
+import net.minecraft.util.Mirror;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.Rotation;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3i;
+import net.minecraft.world.World;
+import net.minecraft.world.gen.feature.template.PlacementSettings;
+import net.minecraft.world.gen.feature.template.Template;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Big shout-outs and credits to Immersive Engineering for providing most of the inspiration and example implementation code for us to figure out this implementation.
```

Comment:
Are you respecting their license if you're pulling that much code that you have to shout out?